### PR TITLE
Support both 0 and null values in a dimension for X axis

### PR DESCRIPTION
**Pull request by alxnddr**
### Description

When numeric dimensions with `null` values were used for X-axis it did not work correctly because Crossfilter merged these keys into a single one `0`. Crossfilter groups were also corrupted when using our null replacement string `(empty)` depending on its order in the dimension.

The solution is hacky, so feel free to suggest better ideas. If we replace `null` with `-Infinity` to perform reduce operation it works correctly. Then we just format it as `(empty)` in axes and tooltips.

#### Linear scale

Before:
Screen Shot 2021-04-09 at 23 15 03

After:
Screen Shot 2021-04-12 at 15 16 43

#### Ordinal scale

Before:
Screen Shot 2021-04-09 at 23 14 23

After:
Screen Shot 2021-04-12 at 15 17 16

#### Original issue with grouping by an hour of day

After:

Data:
Screen Shot 2021-04-12 at 15 18 12

Linear:
Screen Shot 2021-04-12 at 15 18 05

Ordinal:
Screen Shot 2021-04-12 at 15 18 35

Fixes https://github.com/metabase/metabase/issues/12138

### CODE REVIEW

File: frontend/test/metabase/visualizations/lib/renderer_utils.unit.spec.js [hunk 1]
```diff
@@ -123,8 +123,9 @@ describe("getXValues", () => {
       "2019-08-13T00:00:00Z",
     ]);
   });
-  it("should include nulls by default", () => {
-    const xValues = getXValuesForRows([[["foo"], [null], ["bar"]]]);
+  it("should include nulls for ordinal", () => {
```
Comment (alxnddr) [inline frontend/test/metabase/visualizations/lib/renderer_utils.unit.spec.js:126]: @mazameli @flamber is it correct that we should show nulls only on the ordinal axis? I saw many multiple comments in code that say so but the tests expect it to be the default behavior.

**Review by paulrosenzweig — APPROVED** [review]
LGTM! 🎊 🎊 🎊

File: frontend/src/metabase/visualizations/lib/numeric.js [hunk 2]
```diff
@@ -1,7 +1,20 @@
...
+  let hasAtLeastOneNumber = false;
+
+  const hasNumbersOrNullsOnly = rows.every(row => {
+    const isNumber = typeof row[i] === "number";
+    hasAtLeastOneNumber |= isNumber;
+
+    return isNumber || row[i] === null;
+  });
+
+  return hasNumbersOrNullsOnly && hasAtLeastOneNumber;
```
[the remark this answers is not in the document]
Comment (alxnddr) [reply frontend/src/metabase/visualizations/lib/numeric.js]: @paulrosenzweig I agree, it is more readable to have two iterations, so I changed that.  As to why we need this logic, it is necessary for grouping DateTime column by hour of day. The effective type has a higher priority. It looks like a hack, but I'm afraid that changing that might break other places.
Screen Shot 2021-04-12 at 23 14 18

Thanks for the review!

[the remark this answers is not in the document]
Comment (dpsutton) [reply frontend/src/metabase/visualizations/lib/numeric.js]: Some context: the effective_type is there to remove a bunch of edge cases dealing with `semantic_type`. Previously, this column would indicate a bit of semantics around the type: temporal columns could have creation time stamp, updated timestamp, text columns might have name, email, etc, added to them. And then sometimes it meant the data was casted: `:type/UNIXTimestampSeconds` and the like. So nothing on the field said it was a timestamp, it was just an integer field and every codepath that cared about timestamps had to check if the field was some kind of date or see if it's semantic_type caused it to be a date. This is reifying that and making sure that we can always rely on effective_date and not have to know about all of the different ways something might _become_ a temporal type.

A lot of the inference logic hasn't been updated to use effective_type and that's why it's sometimes there and sometimes not. I'm also a bit surprised to see a coercion_strategy is missing when the effective_tyep and base types differ.

[the remark this answers is not in the document]
Comment (alxnddr) [reply frontend/src/metabase/visualizations/lib/numeric.js]: @dpsutton oh, now its purpose is clear. Thanks for sharing this

**Pull request merged**